Check if response has unformatted error details

cpuguy83 opened the pull request:
In some cases a registry may not return a formaatted message, in this
case try to get error details from the response body as a simple string.

Closes #4672

cpuguy83 commented:
Ping @thaJeztah @tonistiigi WDYT?

estesp commented:
Looks like there is some string matching in `TestDockerFetcherOpen` that is broken by this change

cpuguy83 commented:
It like the error handling is just incorrect anyway.
I've actually noticed containerd is always missing error details.

The specific case of 429 returns an object rather than an array:

```
{
  "details": "Too Many Requests. Please see https://docs.docker.com/docker-hub/download-rate-limit/",
  "error": {
    "details": "Too Many Requests. Please see https://docs.docker.com/docker-hub/download-rate-limit/"
  }
}
```

cpuguy83 commented:
But handling the case of this possibly being just an unencoded message is probably OK as well, e.g. maybe the registry is in a degraded state.

estesp commented:
What I'm confused about is that it was "fixed" in #3109 (I'm pretty sure @bainsy88 confirmed at the time that the full error message was getting passed through after that PR), and shouldn't have been broken by #3173, at least on its face, since it is still trying to decode into the same value, but with the more canonical approach of letting the decoder read the stream rather than the "ReadAll(..)".

thaJeztah commented:
Is it because the registry in this case returns `application/json`, but a format that doesn't match?

estesp commented:
Oh, I see Brian's point--the response is not an array, so it's not going to decode into `Errors`; could that have changed at some point with DockerHub? Or maybe the original PR from the IBM registry team still works properly with ICCR/icr.io responses, which may properly be encoded as a JSON array

cpuguy83 commented:
We could update `UnmarshalJSON` for `Errors` to check for this case.

cpuguy83 committed:
docker remote: Fix error unmarshaling

`Errors` here should used lowercase `errors` in the json.

This should fix issues where error details are not present when the
registry returns a non-OK response.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
docker remote: add fallback for error unmarshal

1. See if only a single error is returned if there were no errors in the
   errors list.
2. Fallback to legacy message type from
   https://github.com/docker/distribution/blob/749f6afb4572201e3c37325d0ffedb6f32be8950/registry/client/errors.go#L48-L52

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
docker remote: Coalesce errs for resolve/pull

Makes suere we do the same thing in either case and do our best to get
an error.

Also adds an additional fallback for plain text error responses.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Ok, updated this with some commits split up.

1. The error unmarshalling was incorrect to begin with. It was looking for `Errors` in json instead of `errors`, which explains why I've seen missing error details when using containerd libs.
2. Add fallback to handle:
i. If error is a single error instead of a `{ Errors: []Error }`
ii. Additional fallback for legacy behavior like https://github.com/docker/distribution/blob/749f6afb4572201e3c37325d0ffedb6f32be8950/registry/client/errors.go#L48-L52
3. Coalesce error handling for resolve and fetch, with an additional fallback if the error is plain text.

cpuguy83 commented:
```
$ bin/ctr image pull docker.io/ratelimitalways/test:latest
docker.io/ratelimitalways/test:latest:                                            resolved       |++++++++++++++++++++++++++++++++++++++| 
manifest-sha256:c2d41d2ba6d8b7b4a3ffec621578eb4d9a0909df29dfa2f6fd8a2e5fd0836aed: downloading    |--------------------------------------|    0.0 B/527.0 B 
elapsed: 2.6 s                                                                    total:   0.0 B (0.0 B/s)                                         
ctr: failed to copy: httpReaderSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/ratelimitalways/test/manifests/sha256:c2d41d2ba6d8b7b4a3ffec621578eb4d9a0909df29dfa2f6fd8a2e5fd0836aed: 429 Too Many Requests - Server message: unknown: Too Many Requests. Please see https://docs.docker.com/docker-hub/download-rate-limit/
```

thaJeztah commented:
curious; where is the `unknown` coming from in `Server message: unknown` ?

cpuguy83 commented:
It's from the errcode being 0, since the server is not giving us one in the json.

estesp reviewed (approved):
LGTM

crosbymichael commented:
LGTM but I'll let @dmcgowan do the final check since he knows this better than I do.

dmcgowan commented on remotes/docker/resolver.go:
```diff
@@ -299,7 +299,7 @@ func (r *dockerResolver) Resolve(ctx context.Context, ref string) (string, ocisp
 					continue
 				}
-				return "", ocispec.Descriptor{}, errors.Errorf("unexpected status code %v: %v", u, resp.Status)
+				return "", ocispec.Descriptor{}, getResponseErr(ctx, req, resp)
```
There is a `resp.Body.Close()` a few lines above this

thaJeztah commented:
Should we decide wether or not we must support the invalid JSON response?

thaJeztah commented:
i.e.; I agree that plain text responses should be returned; mostly wondering where the "backward compat" part in moby / distribution came from, and if it's still relevant (not taking the current Docker Hub response into account) https://github.com/containerd/containerd/issues/4672#issuecomment-719003947

dmcgowan commented on remotes/docker/fetcher.go:
```diff
@@ -206,3 +197,23 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
+		if err2 != nil {
+			log.G(ctx).WithError(err2).Debug("Error copying response body details")
+		}
+		return errors.Errorf("unexpected status code %v: %v, details: %s", req.String(), resp.Status, sb)
```
I would rather us type this error, shoveling the raw http body into an error string still doesn't seem right. The cases this occurs is if the registry is not implementing the same error handling format or the response is from middleware. The output is more likely to be mangled html than a valid error message.

We could wrap the registry error type in a response error which holds the request, status code, registry error type. In this case, we can use the unknown registry error type and let the caller decide if it wants to present details.

thaJeztah replied:
> The output is more likely to be mangled html than a valid error message.

Distribution spec is a bit sketchy on that ("If the response body is in JSON format"); i.e. it mentions the body to be JSON format, but doesn't mention wether or not that requires the content-type to be `application/json` (i.e., should a `text/plain` response be attempted to parse as JSON?)

> We could wrap the registry error type in a response error which holds the request, status code, registry error type. In this case, we can use the unknown registry error type and let the caller decide if it wants to present details.

You mean; preserve the body, but convert it to a "valid" format? I think that would work; at least the information would not be _lost_, and clients should be able to handle the standard format (how to handle other formats is somewhat "unspecified":  _"A 4XX response code from the registry MAY return a body in any format"_ )

thaJeztah replied:
@tonistiigi additional thoughts on the above? ^^

cpuguy83 replied:
@dmcgowan That's a fair point. It also kinda sucks to expect all callers to be aware of this to make that determination.
I'd be fine dropping that, or possibly checking explicitly `text/plain` as the content type.

dmcgowan replied:
Can you just dump the body into the log and mention the status code is unexpected and the error is not formatted.

thaJeztah commented:
quick search in the docker/distribution repository; https://github.com/docker/distribution/commit/ec636bbfd246e822bc20c8b1d5e06de965a9469d https://github.com/docker/distribution/pull/1379

> https://github.com/docker/distribution/pull/1249 changed token fetching to parse HTTP error response bodies as serialized errcodes. However, Docker Hub's authentication endpoint does not return error bodies in this format. To work around this, convert its format into ErrCodeUnauthorized or ErrCodeUnknown.

fixes / related to https://github.com/docker/distribution/issues/1373
related to https://github.com/moby/moby/issues/19476

estesp commented:
this PR now has conflicts with `main`; given the last few comments, is there a path forward with @dmcgowan's suggestion of a typed error and/or logging the body content?

dmcgowan commented:
There is still a tracking issue, going to close this for now. We do need to address better remote error response handling still.
